Plugins/BuildServerIntegration/VstsAndTfsIntegration/TfsApiHelper.cs

```diff
@@ -0,0 +1,186 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using GitUI;
+using Microsoft.VisualStudio.Threading;
+using Newtonsoft.Json;
+
+namespace VstsAndTfsIntegration
+{
+    /// <summary>
+    /// For VSTS and TFS >=2015
+    /// </summary>
+    public class TfsApiHelper : IDisposable
+    {
+        private HttpClient _httpClient;
+        private IEnumerable<BuildDefinition> _buildDefinitions;
+        private string _projectId;
+
+        private HttpClient InitializeHttpClient(string serverUrl, string personalAccessToken)
+        {
+            var client = new HttpClient();
+            client.DefaultRequestHeaders.Accept.Add(
+                new MediaTypeWithQualityHeaderValue("application/json"));
+
+            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic",
+                Convert.ToBase64String(
+                    System.Text.Encoding.ASCII.GetBytes($":{personalAccessToken}")));
+
+            var formatedUrl = serverUrl.EndsWith("/") ? serverUrl : serverUrl + "/";
+            client.BaseAddress = new Uri(formatedUrl);
+            return client;
+        }
+
+        private async Task<IEnumerable<BuildDefinition>> GetBuildDefinitionsAsync(string teamCollection, string projectName, Regex buildDefinitionNameFilter)
+        {
+            var projects = await GetProjectsAsync(teamCollection);
+            var buildDefs = new List<BuildDefinition>();
+
+            if (projects.Count == 0)
+            {
+                return buildDefs;
+            }
+
+            var projectFound = projects.FirstOrDefault(p => p.Name == projectName);
+            if (projectFound == null)
+            {
+                return buildDefs;
+            }
+
+            _projectId = projectFound.Id;
+            using (var response = await _httpClient.GetAsync($"{_projectId}/_apis/build/definitions"))
```

Comment:
This call returns all existing build definitions. In some cases, this can take a long time, so it's not really ideal. In the testing I've done, this single request takes ~25-30s, with a payload of ~20MiB. There's several thousands build definitions, so that doesn't help. That request is apparently done each time the commit graph is reloaded, so quite often.

One way to improve the situation would be to leverage the optional [name](https://docs.microsoft.com/en-us/rest/api/vsts/build/definitions/list?view=vsts-rest-4.1) parameter. It supports simple wildcards like *, but not Regex unfortunately.

It would be possible to either make the user enter a build definitions using only wildcard, not regex.
Pro: this is the most optimal solution (perf wise), simple to implement as well
Con: it doesn't match the standard GitExtension system for build definition filtering.

We could also think of a complex logic to approximate the provided Regex with wildcards, and filter the results locally afterward, with the hope that the builds definitions to process would be way smaller.

Another way of tackling the issue would be to see if reusing the list of build definitions across commit graph refresh would be doable. I haven't looked at that approach at all, it might be impractical.

This can be handled in another PR obviously, if you'd rather have that first merged and tested a bit more broadly. I'd be happy to help too :)

Reply:
You can probably cache it, if you could associate it with a commit hash.
The cache can be clearer whenever a different repository is opened.

Reply:
>One way to improve the situation would be to leverage the optional name parameter. It supports simple wildcards like *, but not Regex unfortunately.

I'm ok to stick with the `name` parameter. I'm pretty sure we don't need of the full power of Regex and the * wild char is surely enough.

If we need more, I wonder if we could not use the value to fill the `name` parameter and if no results is returned, call without the parameter and then use it as a regex.

>Another way of tackling the issue would be to see if reusing the list of build definitions across commit graph refresh would be doable. I haven't looked at that approach at all, it might be impractical.

The `BuildServerAdapter` is recreated at each refresh. I found it also surprising and indeed perhaps the creation and the `Initialize()` should be done once. But it's a big refactoring impacting all the plugins... And if we go with that, we should perhaps find a way to trigger this `Initialize()` again if needed.